pkg/mimirpb/split.go

```diff
@@ -37,6 +44,80 @@ func SplitWriteRequestByMaxMarshalSize(req *WriteRequest, reqSize, maxSize int)
 	return merged
 }
 
+func splitWriteRequestByMaxMarshalSizeRW2(req *WriteRequest, reqSize, maxSize int) []*WriteRequest {
+	if len(req.TimeseriesRW2) == 0 {
+		return nil
+	}
+
+	newPartialReq := func() (*WriteRequest, int) {
+		r := &WriteRequest{
+			Source:              req.Source,
+			SkipLabelValidation: req.SkipLabelValidation,
+		}
+
+		return r, r.Size()
+	}
+
+	symbolSize := req.SymbolsRW2Size()
+	timeSeriesSize := req.TimeseriesRW2Size()
+
+	// Naive implementation: send all symbols on all sub-requests.
```

Comment:
I _strongly_ suspect we can't actually do this. Symbols is an exceedingly high proportion of the request, we need to shard it or we can fail to split effectively. Also the duplication cuts into our Kafka throughput.